### PR TITLE
UCT/IB/RC: memory consumption reduction

### PR DESCRIPTION
**Pull request by eladpe**
This PR is based on PR 1409
It cover memory reduction for RC only

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1682/ for details.

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3614/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1684/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3616/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1685/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3617/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1740/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1741/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3685/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3686/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3699/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1761/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1833/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3775/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1835/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3776/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1844/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3785/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3796/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1848/ for details.

**Comment by eladpe** [comment]
@yosefe ready for review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1851/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3807/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_iface.c [hunk 1]
```diff
@@ -156,11 +162,18 @@ ucs_status_t uct_ib_iface_recv_mpool_init(uct_ib_iface_t *iface,
                         config->rx.mp.max_bufs);
     }
 
+    if (!is_accel) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: need to rename the is_accel parameter to express what it does and not how it's used
also, need to calculate grow only if needed, and prefer `if (x) else ` over `if (!x) else` if there is no importance for performance

File: src/uct/ib/rc/base/rc_iface.c [hunk 2]
```diff
@@ -410,9 +412,14 @@ UCS_CLASS_INIT_FUNC(uct_rc_iface_t, uct_rc_iface_ops_t *ops, uct_md_h md,
         goto err;
     }
 
+    if (!strcmp("rc_mlx5",params->tl_name)) {
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: need to pass parameter to indicate how mpool is initialized, no "hacks" please

File: src/uct/ib/rc/base/rc_iface.h [hunk 3]
```diff
@@ -158,7 +158,8 @@ typedef struct uct_rc_iface_ops {
 
 typedef struct uct_rc_srq {
     struct ibv_srq           *srq;
-    unsigned                 available;
+    unsigned                 available; /* Requested number of RX requests */
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: /* Number of buffers which can be posted to HW */

File: src/uct/ib/rc/base/rc_iface.h [hunk 4]
```diff
@@ -158,7 +158,8 @@ typedef struct uct_rc_iface_ops {
...
+    unsigned                 reserved;  /* Reserved number of pre-posted RX requests */
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: /* By how much 'available' can grow */

File: src/uct/ib/rc/verbs/rc_verbs_common.h [hunk 5]
```diff
@@ -163,6 +163,16 @@ uct_rc_verbs_iface_poll_rx_common(uct_rc_iface_t *iface)
                                      wc[i].byte_len, wc[i].imm_data, wc[i].slid);
     }
     iface->rx.srq.available += num_wcs;
+    /* increase the limit of pre-posted RX requests if "DDoS" detected */
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.h]: pls rephrase this comment

File: src/uct/ib/rc/verbs/rc_verbs_common.h [hunk 6]
```diff
@@ -163,6 +163,16 @@ uct_rc_verbs_iface_poll_rx_common(uct_rc_iface_t *iface)
...
+    if ((num_wcs == iface->super.config.rx_max_poll) && iface->rx.srq.reserved) {
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.h]: ucs_unlikely

File: src/uct/ib/ud/base/ud_iface.c [hunk 7]
```diff
@@ -396,8 +396,9 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_ud_iface_ops_t *ops, uct_md_h md,
     ucs_status_t status;
     size_t data_size;
     int mtu;
+    int is_accel = 0;
 
-    ucs_trace_func("%s: iface=%p ops=%p worker=%p rx_headroom=%zu ud_rx_priv_len=%u",
+    ucs_trace_func("%s: iface=%p ops=%p worker=%p rx_headroom=%zu ud_rx_priv_len=%u\n",
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: no need to add another \n in the end of log messages

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1853/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3813/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1854/ for details.

**Comment by eladpe** [comment]
@yosefe ready for review

File: src/uct/ib/dc/base/dc_iface.c [hunk 8]
```diff
@@ -185,7 +185,8 @@ void uct_dc_iface_set_quota(uct_dc_iface_t *iface, uct_dc_iface_config_t *config
 
 UCS_CLASS_INIT_FUNC(uct_dc_iface_t, uct_dc_iface_ops_t *ops, uct_md_h md,
                     uct_worker_h worker, const uct_iface_params_t *params,
-                    unsigned rx_priv_len, uct_dc_iface_config_t *config)
+                    unsigned rx_priv_len, uct_dc_iface_config_t *config,
+                    int is_dc_x)
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: again, parameter name should be according to what it does and not according to how it's used
pls rename this param to "rx_mp_slow_start" or similar

File: src/uct/ib/ud/base/ud_iface.c [hunk 9]
```diff
@@ -390,15 +390,15 @@ void uct_ud_iface_remove_async_handlers(uct_ud_iface_t *iface)
...
                     unsigned ud_rx_priv_len,
-                    const uct_ud_iface_config_t *config)
+                    const uct_ud_iface_config_t *config, int is_ud_x)
 {
     unsigned rx_priv_len, rx_hdr_len;
     ucs_status_t status;
     size_t data_size;
     int mtu;
-    int is_accel = 0;
+    int start_small_chunk = 0; //TODO start small chunk for ud
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: this local variable is not needed

File: src/uct/ib/ud/base/ud_iface.c [hunk 10]
```diff
@@ -456,13 +456,13 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_ud_iface_ops_t *ops, uct_md_h md,
     ucs_ptr_array_init(&self->eps, 0, "ud_eps");
     uct_ud_iface_cep_init(self);
 
-    if (!strcmp("ud_mlx5",params->tl_name)) {
-        is_accel = 1;
+    if (is_ud_x) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: why is ud_x parameter is needed, it does not affect the final decision

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3819/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
@yosefe ready for review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1855/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1856/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3822/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@MattBBaker can you pls post ornl failure?

**Comment by MattBBaker** [comment]
@yosefe Looks like it hanged at udrc/test_ucp_wireup.send_disconnect_onesided_wait/1 with no error message.

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by MattBBaker** [comment]
Still hanging at udrc/test_ucp_wireup.send_disconnect_onesided_wait/1

**Comment by yosefe** [comment]
@MattBBaker did you see this problem before on any other pull request?

**Comment by MattBBaker** [comment]
@yosefe Looking through the Jenkins logs it has been failing every time like this.

**Comment by yosefe** [comment]
only this PR or others as well?

**Comment by eladpe** [comment]
There is also PR for UD only
https://github.com/openucx/ucx/pull/1608

**Comment by yosefe** [comment]
@MattBBaker did you see a hang in udrc/test_ucp_wireup.send_disconnect_onesided_wait/1 in any other PR?

**Comment by MattBBaker** [comment]
@yosefe It looks like it happens in #1608 as well, but none of the others.

**Comment by yosefe** [comment]
@eladpe pls check this

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1900/ for details.

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1901/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1903/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1902/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1904/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3865/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3864/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3863/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3866/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3867/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
@MattBBaker  can you please check if the same problem happens now?
Thanks

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1907/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1908/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3869/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3874/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3875/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
@MattBBaker can you provide the machine spec (lscpu)

**Comment by MattBBaker** [comment]
Requested output below, there is also a page on the wiki that describes some more details on the system.
https://github.com/openucx/ucx/wiki/POD9

```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             2
NUMA node(s):          2
Vendor ID:             AuthenticAMD
CPU family:            16
Model:                 2
Stepping:              3
CPU MHz:               2412.175
BogoMIPS:              4801.98
Virtualization:        AMD-V
L1d cache:             64K
L1i cache:             64K
L2 cache:              512K
L3 cache:              2048K
NUMA node0 CPU(s):     0-3
NUMA node1 CPU(s):     4-7
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3876/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3878/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3880/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
bot:mellanoxlab:retest

**Comment by eladpe** [comment]
BOT:MellanoxLab:retest

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3884/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
BOT:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3885/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3886/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3887/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
@MattBBaker can i upload the problematic code again , and when it hang, can you attach gdb and send me the backtrace.  If it is OK with you tell me when so i will upload it

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3894/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3895/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
@eladpe  I'll grab the new PR and hit it with a gdb backtrace. Unless it's this branch that needs to be debugged?

**Pull request closed without merging**